### ISSUE TITLE
Add some machine-readable output to conan info to improve CI work

### ISSUE THREAD

**Issue by tru** [conan-io/conan#344]
(This discussion started on twitter and is a bit shortened here, sorry for anyone else reading it and think that context is missing).

Given a package I need a printout of all dependencies in the order that they need to be built and the list should only include the _packages_ that are out of date. I am looking for a way from the CI to quickly build a package and all it's dependencies (with test_package to make sure they work) but only if they _need_ to be built.

All information is already available in the info command, it just needs to be formatted in a way so that I can iterate over the packages and build, test and upload them.

So far this is what I have managed to do: https://gist.github.com/08c8a0047f4f5e6d7dbd6d4f00d09f25

This is obviously a PoC and very hacky, but it should demonstrate pretty well what I need it to do.

Questions:
- How to check the package is out of date for the current settings instead of checking just the conanfile.py is out of date.
- Should this be part of the info command (--machine) or something similar or should it be it's own command
- Or should this all be rolled up in the package-tools and even build so you don't have to parse the output of this command in something else and execute the commands.

Looking for input here and something that will benefit more people than just me.

**Comment by memsharded** [comment]
I would say that this is very related to https://github.com/conan-io/conan/issues/308, so I guess @monsdar might be interested in following this issue. Maybe the issues can be even merged.

An important point here is the definition of "out-of-date". In the conan "update" terminology, there is an update of a local package, when an upstream package has been overwritten. But when talking about CI, usually is that the source code has been modified.

So when there is an "outdated" package, what conan needs, is to ``install --update``, but that package does not need to be rebuilt. So conan has no information whether there has been a change in the source code, and the package needs to be rebuilt.

Building a certain package is typically triggered by pushed commits to a repository. Lets try to put an example, to see if we understand the same. 

Example 1:
- Lets say that we have 3 packages that depend: A->B->C.
- If push commits to B, B is automatically packaged, test-package and uploaded the artifact, overwriting the existing one (assuming we are not increasing versions).
- In this context, the package B is already updated locally! In the CI machine. Probably the CI machine should make use of the CONAN_USER_HOME environment variable to guarantee an independent build for each process.
- Now we can ask about the deps of all the packages in our CI, so when we get that C doesn't depend on B, we skip it, and when we get that A depends on B, we re-build A

Example 2:
- Lets say that we have the same 3 packages that depend: A->B->C.
- If push commits to C, C is automatically packaged, test-package and uploaded the artifact, overwriting the existing one
- Now, we can ask for deps of A and B, and both will depend on C. But we have to re-build B first, and doing the analysis at the CI can be painful.

Conan internally already does a sort, of its dependencies, by levels, to process in order the things, that could be put in output in machine readable format. But if we want to define which need to be rebuilt, we need to input the dep that has been modified, and execute over a root (final consumer) package/project.

So in the example, for A->B->C it could be:

```bash
.../folderA $ conan info --build_order_modified=C
B, A
```
In any case, the package which sources has been modified should be provided by the CI system. 
This is slightly different than your proposed, but I think it might be the necessary info. Please feedback, thanks!

**Comment by monsdar** [comment]
This is somehow related to #308, but not exactly the same issue. This is more like automatically resolving dependencies and building dependent packages if a certain package gets updated (automatically building A and B if C gets updated).

In a CI Pipeline the dependencies of components are already configured via the pipeline itself. The CI system knows that A -> B -> C, so if C gets updated it builds B and then builds A. The Pipeline is set up by an integration user.

With a configured CI Pipeline your examples would look like the following:

Example 1 - B gets committed
* User commits B
* CI builds, tests and uploads B
* CI knows that A depends on B
* CI build, tests and uploads A

Example 2 - C gets committed
* User commits C
* CI builds, tests and uploads C
* CI knows that B depends on C
* CI build, tests and uploads B
* CI knows that A depends on B
* CI build, tests and uploads A

Having some type of script that is able to automatically generate dependency trees out of multiple packages sounds quite cool. Making the dependency-information machine-readable would be a first step.

Just a quick idea without having spent much thought on it: This could allow for a mechanism that checks whenever a package is updated on the conan server and then rebuilds all dependent packages that are found on the server. A self-updating conan server... But this and most other things in my post are enough content for multiple additional issues, perhaps we should keep things simple and first integrate the machine-readable output.

**Comment by memsharded** [comment]
Yes, I think this is the idea. Specially because A->B->C is a simple tree, but it happens that projects evolve to have tens or even thousands of packages/libraries, and manually maintaining in the CI the deps tree can be cumbersome and error prone. I will implement an information query to conan that will return an ordered list of package references in the order they have to be built, when a certain package in the dependency tree for the projects is modified. I'll come back soon with a proposal :)

**Comment by raulbocanegra** [comment]
Hi guys,

This is very interesting for CI environments. 
I think that a version update would we useful too.

Imagine the next situation where we need to updates versions:

- C has version 1.5.5.
- B has version 2.3.2.
- A has version 1.0.0

User commits C that has to be updated from 1.5.5 to 2.0.0.
CI builds tests and uploads 2.0.0 version of C.
Then, C version on the requirements of B's conanfile.py should change, and even B version from 2.3.2 to 3.0.0.
Maybe this request should be in another issue entry. I just wanted to let you know.
Thanks for your work!

**Comment by memsharded** [comment]
Hi Raul!
I'd say that this could be a bit problematic. Creating a new version of a package should not produce an automatic upgrade to consumers. I'll put a simple example:
- You have Boost/1.60.0 in your CI, and you have project A that consist of several packages that depend on Boost/1.60.0. You might have also other projects B, C that depend on Boost/1.60.
- Now you are going to start a new project D, and want to try Boost/1.61 for that project. So you bump the version from 1.60 => 1.61 in your Boost package conanfile, test it, and upload it.
- Do you really want CI to automatically upgrade your projects A, B, C to the new Boost version? I don't think so.

I'd say that other package managers, they never do an upgrade downstream, that implies modifying user files in their repositories (as conanfile.txt, or requirements.txt in python or pom.xml in maven...), probably having to commit & push to make that change permanent. We don't want CI or conan, or any other tool, automatically modifying our project configuration or dependencies.

The most used workflow would be:
- Upload the Boost/1.61 package. Nothing changes in projects A, B, C
- Go to project A. Bump version 1.60=>1.61 in the conanfiles. Maybe it is enough to bump in the final conanfile (dependencies are overriden upstreams, so specifying 1.61 in the final consumer, sets 1.61 for the whole project).
- Push project A to CI, so it gets built with the new dependency to Boost/1.61.

A different issue/feature would be to ask for upgrades, to check if there are new versions of your dependencies. That makes sense, although not typically automated in CI either. This would be something that you could run in your machine, get notified of new versions, update if you want, and then push to CI so it builds everything with the new versions. Note that in general it is difficult to apply an automatic policy to accept updates: would you accept an update from channel testing=>stable? would you accept a major version update e.g. 1.2=>2.0? What if the package does not follow semver?

I would like to add my opinion (based on my experience with companies), is that version upgrading is not as fast in C++ as in other languages. Stability is important for the sizes of projects in C++, and upgrading dependencies is always a conscious decision, typically with the consensus of the team. So things as automatic version upgrades or automatically retrieving the latest version of a package sounds like cool features for a package manager but are rarely used in practice in C/C++ production environments.

In any case, if you still want to automatically bump versions in consumer packages in your CI, we can try to help providing information features, or helping to create hooks in the conan_server that allows you to automate such task in your CI. Like the previously proposed one, for example, I am almost done, I think tomorrow I will be able to send a PR :)

Thanks very much for your feedback!

**Comment by memsharded** [comment]
I have submitted this PR proposal: https://github.com/conan-io/conan/pull/358

It works as:

```bash
$ conan info --build_order=Zlib/0.1@user/channel
[Zlib/0.1@user/channel], [Poco/0.1@user/channel, Boost/0.1@user/channel], [PROJECT]
```
For a given project, it provides the dependency that has been modified (pushed to CI. It allows also multiple dependencies).
It returns an ordered list of lists in the required build order. Internal lists are independent packages, so CI can build them in parallel, once the previous levels have been built.

It manages scopes, i.e. if a ``dev_requires`` is modified, it returns empty, unless the correspondent ``dev`` scope is set to True.

It also allows to retrieve the full build order (it manages scopes too):
```bash
$ conan info --build_order=ALL
... full build order for all projects
```

I hope this feature will be useful for CI, feedback welcome

**Comment by raulbocanegra** [comment]
Hi James,

I know what you mean and you are right. That will be a mess.
In my example above I was thinking about internal libraries. So I would like to clarify my point of view.
I'm currently working on a legacy project with several internal libraries. By internal libraries I mean libraries that are part of our legacy project which are only maintained by our team, I didn´t  mean Boost, Poco or other "external" libraries. 
As a legacy, we have to face some problems, one of them is that libraries are very coupled, so a changed from 1.Y.Z to 2.Y.Z makes a lot of changes in the version of the rest the project libraries.
I thought this would be something that Conan would address until we have some time to refactor our legacy libraries.

Thanks for your work!

**Comment by tru** [comment]
Hey! finally circling back to this. I think this looks really good @memsharded - the only missing part from my pov is to have a way to see if the package "needs" to be rebuilt not checked against the local repo but the remote one.

**Comment by memsharded** [comment]
Hi! Thank you!
Ok, but I don't understand. The local and remote repos are git repositories that have been modified? Can you elaborate a bit more the use case? Thanks!

**Comment by tru** [comment]
Alright here is my usecase:

One repository with several conanfiles in subdirectories:
```
libxml2/conanfile.py
bzip2/conanfile.py
...
```

Then I also have a "meta" conanfile to build all dependencies (in the future this conanfile is probably going to be auto-generated to get all the latest versions so we don't have to update that manually).

There is then a CI job that builds this meta package. With the output above I can actually print the order to build them and then run `test_package` on each of them and then upload them.

Now we have a stable base where all packages on the remote are up-to-date. Now ponder we want to update the libxml2 package. With the current setup it would require us to rebuild *all* packages which will be a very expensive and slow process. So I would ideally want to know if the package *needs* to be rebuild by checking the remote.

There will always be a need for a override switch in the CI to rebuild all packages when we change something that breaks the ABI.

Is that explaining it correctly?

**Comment by memsharded** [comment]
The output of conan commands as machine readable is followed in modern issue: https://github.com/conan-io/conan/issues/1895
